asmcli/lib/util.sh

```diff
@@ -22,3 +22,978 @@ gen_install_params() {
 
   echo "${PARAM_BUILDER}"
 }
+
+#######
+# run_command takes a list of arguments that represents a command
+# If DRY_RUN or VERBOSE is enabled, it will print the command, and if DRY_RUN is
+# not enabled it runs the command.
+#######
+run_command() {
+  local DRY_RUN; DRY_RUN="$(context_get-option "DRY_RUN")"
+  local VERBOSE; VERBOSE="$(context_get-option "VERBOSE")"
+
+  if [[ "${DRY_RUN}" -eq 1 ]]; then
+    warn "Would have executed: ${*}"
+    return
+  elif [[ "${VERBOSE}" -eq 0 ]]; then
+    "${@}" 2>/dev/null
+    return "$?"
+  fi
+  info "Running: '${*}'"
+  info "-------------"
+  local RETVAL
+  { "${@}"; RETVAL="$?"; } || true
+  return $RETVAL
+}
+
+#######
+# retry takes an integer N as the first argument, and a list of arguments
+# representing a command afterwards. It will retry the given command up to N
+# times before returning 1. If the command is kubectl, it will try to
+# re-get credentials in case something caused the k8s IP to change.
+#######
+retry() {
+  local MAX_TRIES; MAX_TRIES="${1}";
+  shift 1
+  for i in $(seq 0 "${MAX_TRIES}"); do
+    if [[ "${i}" -eq "${MAX_TRIES}" ]]; then
+      break
+    fi
+    { "${@}" && return 0; } || true
+    warn "Failed, retrying...($((i+1)) of ${MAX_TRIES})"
+    sleep 2
+  done
+  local CMD="'$*'"
+  warn "Command $CMD failed."
+  false
+}
+
+find_missing_strings() {
+  local NEEDLES; NEEDLES="${1}";
+  local HAYSTACK; HAYSTACK="${2}";
+  local NOTFOUND; NOTFOUND="";
+
+  while read -r needle; do
+    EXITCODE=0
+    grep -q "${needle}" <<EOF || EXITCODE=$?
+${HAYSTACK}
+EOF
+    if [[ "${EXITCODE}" -ne 0 ]]; then
+      NOTFOUND="${needle},${NOTFOUND}"
+    fi
+  done <<EOF
+${NEEDLES}
+EOF
+
+  if [[ -n "${NOTFOUND}" ]]; then NOTFOUND="$(strip_trailing_commas "${NOTFOUND}")"; fi
+  echo "${NOTFOUND}"
+}
+
+strip_trailing_commas() {
+  # shellcheck disable=SC2001
+  echo "${1}" | sed 's/,*$//g'
+}
+
+warn() {
+  info "[WARNING]: ${1}" >&2
+}
+
+warn_pause() {
+  warn "${1}"
+  sleep 2
+}
+
+error() {
+  info "[ERROR]: ${1}" >&2
+}
+
+info() {
+  local VERBOSE; VERBOSE="$(context_get-option "VERBOSE")"
+  if hash ts 2>/dev/null && [[ "${VERBOSE}" -eq 1 ]]; then
+    echo "${SCRIPT_NAME}: ${1}" | TZ=utc ts '%Y-%m-%dT%.T' >&2
+  else
+    echo "${SCRIPT_NAME}: ${1}" >&2
+  fi
+}
+
+fatal() {
+  error "${1}"
+  exit 2
+}
+
+fatal_with_usage() {
+  error "${1}"
+  usage_short >&2
+  exit 2
+}
+
+prompt_user_for_value() {
+  read -r -p "Please provide a value for ${1}:" VALUE
+  if [[ -n "${VALUE}" ]]; then
+    echo "${VALUE}"
+  fi
+}
+
+starline() {
+  echo "*****************************"
+}
+
+enable_common_message() {
+  echo "Alternatively, use --enable_all|-e to allow this tool to handle all dependencies."
+}
+
+arg_required() {
+  if [[ ! "${2:-}" || "${2:0:1}" = '-' ]]; then
+    fatal "Option ${1} requires an argument."
+  fi
+}
+
+parse_args() {
```

Comment:
I think arg parsing functions probably don't belong in util.sh. Especially since we're adding `experimental` etc. and subcommands might have different parsing logic, we should probably have a dedicated file for shared arg parsing/validation logic.

Reply:
Done.